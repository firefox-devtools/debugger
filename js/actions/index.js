const breakpoints = require('./breakpoints');
const eventListeners = require('./event-listeners');
const sources = require('./sources');
const tabs = require('./tabs');

module.exports = Object.assign({}, breakpoints, eventListeners, sources, tabs);
