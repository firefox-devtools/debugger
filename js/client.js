const { DebuggerClient } = require('devtools/shared/client/main');
const { DebuggerTransport } = require('devtools/transport/transport');
const { TargetFactory } = require('devtools/client/framework/target');
const promise = require('devtools/sham/promise');

const socket = new WebSocket('ws://localhost:9000');
const transport = new DebuggerTransport(socket);
const client = new DebuggerClient(transport);

function connectToClient(onConnect) {
  client.connect().then(() => {
    return client.listTabs().then(onConnect);
  }).catch(err => console.log(err));
}

function connectToTab(tab, onNewSource) {
  let deferred = promise.defer();
  const options = { client, form: tab, chrome: false };

  TargetFactory.forRemoteTab(options).then(target => {
    target.activeTab.attachThread({}, (res, threadClient) => {
      threadClient.resume();
      window.gThreadClient = threadClient;
      deferred.resolve();
    });
  });

  return deferred.promise;
}

module.exports = {
  connectToClient,
  connectToTab
};
