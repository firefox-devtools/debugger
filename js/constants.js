/* -*- indent-tabs-mode: nil; js-indent-level: 2 -*- */
/* vim: set ft=javascript ts=2 et sw=2 tw=80: */
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
'use strict';

exports.UPDATE_EVENT_BREAKPOINTS = 'UPDATE_EVENT_BREAKPOINTS';
exports.FETCH_EVENT_LISTENERS = 'FETCH_EVENT_LISTENERS';

exports.TOGGLE_PRETTY_PRINT = 'TOGGLE_PRETTY_PRINT';
exports.BLACKBOX = 'BLACKBOX';

exports.ADD_BREAKPOINT = 'ADD_BREAKPOINT';
exports.REMOVE_BREAKPOINT = 'REMOVE_BREAKPOINT';
exports.ENABLE_BREAKPOINT = 'ENABLE_BREAKPOINT';
exports.DISABLE_BREAKPOINT = 'DISABLE_BREAKPOINT';
exports.SET_BREAKPOINT_CONDITION = 'SET_BREAKPOINT_CONDITION';

exports.ADD_SOURCE = 'ADD_SOURCE';
exports.LOAD_SOURCES = 'LOAD_SOURCES';
exports.LOAD_SOURCE_TEXT = 'LOAD_SOURCE_TEXT';
exports.SELECT_SOURCE = 'SELECT_SOURCE';
exports.UNLOAD = 'UNLOAD';
exports.RELOAD = 'RELOAD';

exports.ADD_TABS = 'ADD_TABS';
exports.SELECT_TAB = 'SELECT_TAB';
