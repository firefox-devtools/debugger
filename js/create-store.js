/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
'use strict';

const { createStore, applyMiddleware } = require('redux');
const { thunk } = require('devtools/client/shared/redux/middleware/thunk');
const { waitUntilService } = require('devtools/client/shared/redux/middleware/wait-service');
const { log } = require('devtools/client/shared/redux/middleware/log');
const { promise } = require('devtools/client/shared/redux/middleware/promise');
const { history } = require('devtools/client/shared/redux/middleware/history');

/**
 * This creates a dispatcher with all the standard middleware in place
 * that all code requires. It can also be optionally configured in
 * various ways, such as logging and recording.
 *
 * @param {object} opts:
 *        - log: log all dispatched actions to console
 *        - history: an array to store every action in. Should only be
 *                   used in tests.
 *        - middleware: array of middleware to be included in the redux store
 */
module.exports = (opts = {}) => {
  const middleware = [
    thunk,
    promise,

    // Order is important: services must go last as they always
    // operate on "already transformed" actions. Actions going through
    // them shouldn't have any special fields like promises, they
    // should just be normal JSON objects.
    waitUntilService
  ];

  if (opts.history) {
    middleware.push(history(opts.history));
  }

  if (opts.middleware) {
    opts.middleware.forEach(fn => middleware.push(fn));
  }

  if (opts.log) {
    middleware.push(log);
  }

  return applyMiddleware(...middleware)(createStore);
};
