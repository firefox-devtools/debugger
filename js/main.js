const React = require('react');
const ReactDOM = require('react-dom');
const { combineReducers, applyMiddleware, bindActionCreators } = require('redux');
const { Provider } = require('react-redux');
const configureStore = require('./create-store');
const reducers = require('./reducers');
const { connectToClient } = require('./client');
const actions = require('./actions');
const dom = React.DOM;
const TabList = React.createFactory(require('./components/TabList'));

const createStore = configureStore({ log: true });
const store = createStore(combineReducers(reducers));
window.store = store; // global for debugging purposes only!

connectToClient(response => {
  const tabs = response.tabs;
  store.dispatch(actions.newTabs(response.tabs));

  // if there's a pre-selected tab, connect to it and load the sources.
  // otherwise, just show the toolbox.
  if (hasSelectedTab()) {
    const selectedTab = getSelectedTab(tabs);
    store.dispatch(actions.selectTab({ tabActor: selectedTab.actor }))
      .then(() => store.dispatch(actions.loadSources()))
      .then(renderToolbox);
  } else {
    renderToolbox();
  }
});

/**
 * Check to see if the url hash has a selected tab
 * e.g. #tab=child2
 */
function hasSelectedTab() {
  return window.location.hash.includes('tab');
}

/**
 * get the selected tab from the url hash
 * e.g. #tab=child2
 *
 * tabs are keyed by their child id,
 * this is because the actor connection id increments every refresh and
 * tab id is always 1.
 *
 */
function getSelectedTab(tabs) {
  const childId = window.location.hash.split('=')[1];
  return tabs.find((tab) => tab.actor.includes(childId));
}

function renderToolbox() {
  ReactDOM.render(
    React.createElement(
      Provider,
      { store },
      TabList()
    ),
    document.querySelector('#mount')
  );
}
