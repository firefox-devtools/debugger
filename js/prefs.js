const { PrefsHelper } = require('devtools/client/shared/prefs');

/**
 * Shortcuts for accessing various debugger preferences.
 */
let Prefs = new PrefsHelper('devtools', {
  workersAndSourcesWidth: ['Int', 'debugger.ui.panes-workers-and-sources-width'],
  instrumentsWidth: ['Int', 'debugger.ui.panes-instruments-width'],
  panesVisibleOnStartup: ['Bool', 'debugger.ui.panes-visible-on-startup'],
  variablesSortingEnabled: ['Bool', 'debugger.ui.variables-sorting-enabled'],
  variablesOnlyEnumVisible: ['Bool', 'debugger.ui.variables-only-enum-visible'],
  variablesSearchboxVisible: ['Bool', 'debugger.ui.variables-searchbox-visible'],
  pauseOnExceptions: ['Bool', 'debugger.pause-on-exceptions'],
  ignoreCaughtExceptions: ['Bool', 'debugger.ignore-caught-exceptions'],
  sourceMapsEnabled: ['Bool', 'debugger.source-maps-enabled'],
  prettyPrintEnabled: ['Bool', 'debugger.pretty-print-enabled'],
  autoPrettyPrint: ['Bool', 'debugger.auto-pretty-print'],
  workersEnabled: ['Bool', 'debugger.workers'],
  editorTabSize: ['Int', 'editor.tabsize'],
  autoBlackBox: ['Bool', 'debugger.auto-black-box'],
  promiseDebuggerEnabled: ['Bool', 'debugger.promise']
});

module.exports = Prefs;
