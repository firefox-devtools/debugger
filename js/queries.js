
function getSource(state, actor) {
  return state.sources.sources[actor];
}

function getSources(state) {
  return state.sources.sources;
}

function getSourceCount(state) {
  return Object.keys(state.sources.sources).length;
}

function getSourceByURL(state, url) {
  for (let k in state.sources.sources) {
    const source = state.sources.sources[k];
    if (source.url === url) {
      return source;
    }
  }
}

function getSourceByActor(state, actor) {
  for (let k in state.sources.sources) {
    const source = state.sources.sources[k];
    if (source.actor === actor) {
      return source;
    }
  }
}

function getSelectedSource(state) {
  return state.sources.sources[state.sources.selectedSource];
}

function getSelectedSourceOpts(state) {
  return state.sources.selectedSourceOpts;
}

function getSourceText(state, actor) {
  return state.sources.sourcesText[actor];
}

function getBreakpoints(state) {
  return Object.keys(state.breakpoints.breakpoints).map(k => {
    return state.breakpoints.breakpoints[k];
  });
}

function getBreakpoint(state, location) {
  return state.breakpoints.breakpoints[makeLocationId(location)];
}

function getTabs(state) {
  return state.tabs.tabs;
}

function getSelectedTab(state) {
  return state.tabs.selectedTab;
}

function makeLocationId(location) {
  return location.actor + ':' + location.line.toString();
}

module.exports = {
  getSource,
  getSources,
  getSourceCount,
  getSourceByURL,
  getSourceByActor,
  getSelectedSource,
  getSelectedSourceOpts,
  getSourceText,
  getBreakpoint,
  getBreakpoints,
  getTabs,
  getSelectedTab,
  makeLocationId
};
