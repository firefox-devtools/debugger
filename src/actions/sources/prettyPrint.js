/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import assert from "../../utils/assert";
import { recordEvent } from "../../utils/telemetry";
import { remapBreakpoints } from "../breakpoints";

import { setPausePoints, setSymbols } from "../ast";
import { prettyPrint } from "../../workers/pretty-print";
import { setSource } from "../../workers/parser";
import { getPrettySourceURL, isLoaded } from "../../utils/source";
import { loadSourceText } from "./loadSourceText";
import { mapFrames } from "../pause";
import { selectSpecificLocation } from "../sources";

import {
  getSource,
  getSourceFromId,
  getSourceByURL,
  getSelectedLocation
} from "../../selectors";

import type { Action, ThunkArgs } from "../types";
import { selectSource } from "./select";
import type { JsSource } from "../../types";

export function createPrettySource(sourceId: string) {
  return async ({ dispatch, getState, sourceMaps }: ThunkArgs) => {
    const source = getSourceFromId(getState(), sourceId);
    const url = getPrettySourceURL(source.url);
    const id = await sourceMaps.generatedToOriginalId(sourceId, url);

    const prettySource: JsSource = {
      url,
      relativeUrl: url,
      id,
      isBlackBoxed: false,
      isPrettyPrinted: true,
      isWasm: false,
      contentType: "text/javascript",
      loadedState: "loading"
    };

    dispatch(({ type: "ADD_SOURCE", source: prettySource }: Action));
    dispatch(selectSource(prettySource.id));

    const { code, mappings } = await prettyPrint({ source, url });
    await sourceMaps.applySourceMap(source.id, url, code, mappings);

    const loadedPrettySource: JsSource = {
      ...prettySource,
      text: code,
      loadedState: "loaded"
    };

    setSource(loadedPrettySource);

    dispatch(({ type: "UPDATE_SOURCE", source: loadedPrettySource }: Action));

    return prettySource;
  };
}

/**
 * Toggle the pretty printing of a source's text. All subsequent calls to
 * |getText| will return the pretty-toggled text. Nothing will happen for
 * non-javascript files.
 *
 * @memberof actions/sources
 * @static
 * @param string id The source form from the RDP.
 * @returns Promise
 *          A promise that resolves to [aSource, prettyText] or rejects to
 *          [aSource, error].
 */
export function togglePrettyPrint(sourceId: string) {
  return async ({ dispatch, getState, client, sourceMaps }: ThunkArgs) => {
    const source = getSource(getState(), sourceId);
    if (!source) {
      return {};
    }

    if (!source.isPrettyPrinted) {
      recordEvent("pretty_print");
    }

    if (!isLoaded(source)) {
      await dispatch(loadSourceText(source));
    }

    assert(
      sourceMaps.isGeneratedId(sourceId),
      "Pretty-printing only allowed on generated sources"
    );

    const selectedLocation = getSelectedLocation(getState());
    const url = getPrettySourceURL(source.url);
    const prettySource = getSourceByURL(getState(), url);

    const options = {};
    if (selectedLocation) {
      options.location = await sourceMaps.getOriginalLocation(selectedLocation);
    }

    if (prettySource) {
      const _sourceId = prettySource.id;
      return dispatch(
        selectSpecificLocation({ ...options.location, sourceId: _sourceId })
      );
    }

    const newPrettySource = await dispatch(createPrettySource(sourceId));

    await dispatch(remapBreakpoints(sourceId));
    await dispatch(mapFrames());
    await dispatch(setPausePoints(newPrettySource.id));
    await dispatch(setSymbols(newPrettySource.id));

    dispatch(
      selectSpecificLocation({
        ...options.location,
        sourceId: newPrettySource.id
      })
    );

    return newPrettySource;
  };
}
