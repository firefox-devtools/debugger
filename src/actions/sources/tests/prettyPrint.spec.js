/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import {
  actions,
  selectors,
  createStore,
  makeSource
} from "../../../utils/test-head";
import { createPrettySource } from "../prettyPrint";
import { sourceThreadClient } from "../../tests/helpers/threadClient.js";

describe("sources - pretty print", () => {
  const { dispatch, getState } = createStore(sourceThreadClient);

  it("returns a pretty source for a minified file", async () => {
    const url = "base.js";
    const csr = makeSource(url);
    await dispatch(actions.newSource(csr));
    await dispatch(createPrettySource(csr.source.id));

    const prettyURL = `${csr.source.url}:formatted`;
    const pretty = selectors.getSourceByURL(getState(), prettyURL);
    expect(pretty && pretty.contentType).toEqual("text/javascript");
    expect(pretty && pretty.url.includes(prettyURL)).toEqual(true);
    expect(pretty).toMatchSnapshot();
  });

  it("should create a source when first toggling pretty print", async () => {
    const csr = makeSource("foobar.js", { loadedState: "loaded" });
    await dispatch(actions.togglePrettyPrint(csr.source.id));
    expect(selectors.getSourceCount(getState())).toEqual(2);
  });

  it("should not make a second source when toggling pretty print", async () => {
    const csr = makeSource("foobar.js", { loadedState: "loaded" });
    await dispatch(actions.togglePrettyPrint(csr.source.id));
    expect(selectors.getSourceCount(getState())).toEqual(2);
    await dispatch(actions.togglePrettyPrint(csr.source.id));
    expect(selectors.getSourceCount(getState())).toEqual(2);
  });
});
