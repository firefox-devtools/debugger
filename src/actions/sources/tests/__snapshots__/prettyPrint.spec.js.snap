// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sources - pretty print returns a pretty source for a minified file 1`] = `
Object {
  "contentType": "text/javascript",
  "error": undefined,
  "id": "base.js/originalSource-36c718d4bde9a75edb388ff7733efe7f",
  "isBlackBoxed": false,
  "isPrettyPrinted": true,
  "isWasm": false,
  "loadedState": "loaded",
  "relativeUrl": "http://localhost:8000/examples/base.js:formatted",
  "sourceMapURL": undefined,
  "text": "undefined
",
  "thread": "",
  "url": "http://localhost:8000/examples/base.js:formatted",
}
`;
